hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java

```diff
@@ -0,0 +1,487 @@
+        ? extends Table.KeyValue<String, T>> tableIterator,
+            String prefixKey, String startKey,
+            Set<String> cacheDeletedKeySet) throws IOException {
+      this.iterType = iterType;
+      this.tableIterator = tableIterator;
+      this.prefixKey = prefixKey;
+      this.cacheDeletedKeySet = cacheDeletedKeySet;
+      this.currentKey = null;
+      if (!StringUtils.isBlank(prefixKey)) {
+        tableIterator.seek(prefixKey);
+      }
+
+      if (!StringUtils.isBlank(startKey)) {
+        tableIterator.seek(startKey);
+      }
+
+      getNextKey();
+    }
+
+    private void getNextKey() throws IOException {
+      while (tableIterator.hasNext() && currentKey == null) {
+        Table.KeyValue<String, T> entry = tableIterator.next();
+        String entryKey = entry.getKey();
+        if (entryKey.startsWith(prefixKey)) {
+          if (!cacheDeletedKeySet.contains(entryKey)) {
+            currentKey = new HeapEntry(iterType, entryKey, entry.getValue());
+          }
+        } else {
+          break;
+        }
+      }
+    }
+
+    public boolean hasNext() {
+      try {
+        getNextKey();
+      } catch (Throwable t) {
+        throw new NoSuchElementException();
+      }
+      return currentKey != null;
+    }
+
+    public HeapEntry next() {
+      try {
+        getNextKey();
+      } catch (Throwable t) {
+        throw new NoSuchElementException();
+      }
+      HeapEntry ret = currentKey;
+      currentKey = null;
+
+      return ret;
+    }
+
+    public void close() {
+
+    }
+  }
+
+  private static class CacheIter< T extends WithParentObjectId>
+      implements OzoneKeyIterator {
+    private final Set<String> cacheDeletedKeySet;
+    private final Map<String, T> cacheKeyMap;
+
+    private final Iterator<Map.Entry<String, T>>
+        cacheCreatedKeyIter;
+
+    private final Iterator<Map.Entry<CacheKey<String>, CacheValue<T>>>
+        cacheIter;
+
+    private final String prefixKey;
+    private final String startKey;
+
+
+    private final EntryType entryType;
+
+    public CacheIter(EntryType entryType, Iterator<Map.Entry<CacheKey<String>,
+        CacheValue<T>>> cacheIter, String startKey, String prefixKey) {
+      this.cacheDeletedKeySet = new TreeSet<>();
+      this.cacheKeyMap = new TreeMap<>();
+
+      this.cacheIter = cacheIter;
+      this.startKey = startKey;
+      this.prefixKey = prefixKey;
+      this.entryType = entryType;
+
+      getCacheValues();
+
+      cacheCreatedKeyIter = cacheKeyMap.entrySet().iterator();
+    }
+
+    private void getCacheValues() {
+      while (cacheIter.hasNext()) {
+        Map.Entry<CacheKey<String>, CacheValue<T>> entry =
+            cacheIter.next();
+        String cacheKey = entry.getKey().getCacheKey();
+        T cacheOmInfo = entry.getValue().getCacheValue();
+        // cacheOmKeyInfo is null if an entry is deleted in cache
+        if (cacheOmInfo == null) {
+          cacheDeletedKeySet.add(cacheKey);
```

Comment:
Will this run into the same issue as the other commands that needed to be fixed where the memory can grow unbounded?

Reply:
done, using existing method in KeyManagerImpl to check for deleted keys